packages/desktop-client/src/components/accounts/Account.tsx

```diff
@@ -175,208 +440,113 @@ function AllTransactions({
     splitsExpandedDispatch,
   ]);
 
-  transactions ??= [];
+  const calculateRunningBalances = useCallback(async () => {
+    const {
+      data: balances,
+    }: { data: Array<{ id: TransactionEntity['id']; balance: number }> } =
+      await runQuery(
+        queries
+          .transactions(accountId)
+          .options({ splits: 'none' })
+          .orderBy({ date: 'desc' })
+          .select([{ balance: { $sumOver: '$amount' } }]),
+      );
 
-  let runningBalance = useMemo(() => {
-    if (!showBalances) {
-      return 0;
-    }
+    const latestBalance = balances[0]?.balance ?? 0;
 
-    return balances && transactions?.length > 0
-      ? (balances[transactions[0].id]?.balance ?? 0)
-      : 0;
-  }, [showBalances, balances, transactions]);
+    const nonChildPreviewTransactions = previewTransactions.filter(
+      t => !t.is_child,
+    );
 
-  const prependBalances = useMemo(() => {
-    if (!showBalances) {
-      return null;
-    }
+    const nonChildPreviewBalancesById: typeof runningBalances =
+      nonChildPreviewTransactions.reduce(
+        (map, trans, index, array) => {
+          map[trans.id] = {
+            balance: array
+              .slice(index, array.length)
+              .reduce(
+                (sum, t) => sum + getScheduledAmount(t.amount),
+                latestBalance,
+              ),
+          };
+          return map;
+        },
+        {} as Record<string, { balance: number }>,
+      );
+    const balancesById = groupById<{ id: string; balance: number }>(balances);
 
-    // Reverse so we can calculate from earliest upcoming schedule.
-    const previewBalances = [...previewTransactions]
-      .reverse()
-      .map(previewTransaction => {
-        return {
-          // TODO: fix me
-          // eslint-disable-next-line react-hooks/exhaustive-deps
-          balance: (runningBalance += previewTransaction.amount),
-          id: previewTransaction.id,
-        };
-      });
-    return groupById(previewBalances);
-  }, [showBalances, previewTransactions, runningBalance]);
+    return {
+      ...nonChildPreviewBalancesById,
+      ...balancesById,
+    } as typeof runningBalances;
+  }, [accountId, previewTransactions]);
 
-  const allTransactions = useMemo(() => {
-    // Don't prepend scheduled transactions if we are filtering
-    if (!filtered && previewTransactions.length > 0) {
-      return previewTransactions.concat(transactions);
-    }
-    return transactions;
-  }, [filtered, previewTransactions, transactions]);
+  useEffect(() => {
+    let isUnmounted = false;
```

Comment:
I think this is more readable
```suggestion
    let isMounted = true;
```